### ISSUE TITLE
World-specific settings

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#6966]
##### Issue type
 - Feature request
 
##### Summary
Currently, most settings get mangled together in a single global `minetest.conf`. But you quickly run into problems when there's a setting for which different values for different worlds make sense, rather than a globally shared single value.

The most striking example is `time_speed`. I have to always manually change this between 0 and 72 when I switch between development worlds (where I don't want the day/night cycle) and playing worlds (where I want it).

Examples for candidates for world-speficic settings:
* `time_speed`
* `enable_pvp`
* `enable_damage`
* Mod stuff: Disable mob spawning, disable weather changes, disable fire, disable TNT, …

I suggest to add a new setting category: “World-specific setting”. This is a setting which is set on a per-world basis. This would allow me to have a different `time_speed` for each world and I never have to bother about it again.

Minetest already implements this to some extent, but there are holes in the system. There is no standard (!) way to add a custom world-specific setting. There is no server command either.

Deprecates https://github.com/minetest/minetest/issues/5353.

**Comment by ThomasMonroe314** [comment]
Actually, I think a better idea is to make them global unless it is explicitly world-specific, that way it reduces the number of lines in the .conf per-world. They could work in a CSS type fashion, where the last .conf checked has the highest priority. Checking the minetest.conf then the world.conf would mean that the world.conf could override settings in the minetest.conf.

**Comment by paramat** [comment]
Neglected PR #6528 
Related #5353

**Comment by red-001** [comment]
working on a replacement PR that approaches the issue from a slightly different angle, meaning that it wouldn't separate client and server settings but should be a lot easier to review and merge. Rebasing the old PR would be as much work as writing it in the first place.

**Comment by red-001** [comment]
See #6966

**Comment by lhofhansl** [comment]
I strongly support this.

It seems we had multiple attempts that aborted either due to complexity or too many discussions...?
I suppose one of the hardest parts is to have a UI that makes sense and allows for both world and global settings.

**Comment by red-001** [comment]
Yeah the UI is the hardest part to get right, would make some changes to the C++ part if I was to work on it again but the UI was the biggest issue.

**Comment by srutzky** [comment]
I, too, very much support this suggestion, and would love to see it get implemented. I have two questions related to this:

1. I see from the previous couple of comments that the main blocker appears to be getting the UI right to handle the new concept. That's understandable. However, I was wondering if the UI alone should really hold things up. Can't this be done in a phased approach, with the initial concept rolling out as "manual" only (similar to how the left and right mouse buttons are now configurable in Minetest 5.4, but only via the **minetest.conf** file)? I do realize that this is easier if the direction taken for global settings is that they largely remain the same with per-world settings being overrides. But, even if there's a shift away from global settings and pushing most things to the per-world config file, there's still a place for the global settings being a default template for new worlds (at the very least). It's not perfect on initial rollout, but it's progress, and I (and I'm sure many others) can live without the convenience of the UI (for configuring) much more than I / we can live without the ability to (easily) have vastly different options per world, across many worlds (hence manually updating **minetest.conf** isn't feasible).

2. Speaking of incremental improvement, I found something that get's us very close to this goal (even if it wasn't meant to) and that doesn't appear to have been mentioned yet. It is the `minetest.get_mapgen_setting()` function, which allows values found in the **map_meta.txt** file (found in each world directory) to override whatever was found in the **minetest.conf** file. I have tested this in a few mods by replacing several calls to `minetest.settings:get()` with `minetest.get_mapgen_setting()`, and it appears to work fairly well.

    One issue _was_ that, in Minetest 5.3, exiting the game would (always?) overwrite the **map_meta.txt** file, but with only the config options it knows of. Meaning, you could edit the **map_meta.txt** file to include mod options that would be honored as expected (assuming you also updated the mod to use the `minetest.get_mapgen_setting()` function), but as soon as you exit the game the options would no longer be in the file. That behavior clearly made this approach infeasible. However, as luck would have it, a change was made in 5.4 (possibly in this [commit](https://github.com/minetest/minetest/commit/37a05ec8d6cbf9ff4432225cffe78c16fdd0647d#diff-742a716324b97c7b673cb41a50625f5583e55761b4cd8bacaaadf2b2480d0050)) and the **map_meta.txt** file is no longer written out each time you exit the game. Hence, manually added options no longer disappear.

    Now, I do realize that the `minetest.get_mapgen_setting()` function was not _intended_ for this purpose (i.e. general per-world settings), but it appears to at least be _suitable_ for this purpose. From what I can see, it looks to be 80% - 90% of what is desired, right? Of course, the game doesn't look in this file for built-in options that aren't for mapgen (e.g. I tried putting `connected_glass` in **map_meta.txt** and it doesn't apply that as an override), so that would have to change, but if that's the only change, then perhaps `minetest.get_mapgen_setting()` can be renamed to something like `get_world_setting` (but I guess not exactly that as it would be misleading with regards to the **world.mt** file)?

    Again, I realize that this isn't ideal, but it might be a lower-cost approach so I thought I would mention it. For the time being, as of Minetest 5.4 (thank you, thank you, thank you), even though I need to edit some mods to swap out the function used to get certain settings, I can at least now have per-world settings for things like cloudlands, nether, magma conduits, etc, to go along with the different `mapgen_limit` and `water_level` values (which were already supported via the **map_meta.txt** file).

**Comment by SmallJoker** [comment]
It was definitely not the idea of 37a05ec8d6 to allow world-specific settings, but to add more setting "layers" here:
https://github.com/minetest/minetest/blob/c9eba8440d3dc293a8aa6ffafc045737732da1e1/src/settings.h#L63-L69

You could (in theory) use the mapgen settings, but it's bad design, undocumented and not guaranteed to work in first place.
What's needed is basically only some code to:
1. Load a world-specific configuration file into another layer
2. Sanitize the settings (i.e. disallow/strip "secure" settings)
3. A script API that can be used to read out the settings (preferably similar to `minetest.settings` rather than custom `get_mapgen_setting`-style functions)

**Comment by srutzky** [comment]
@SmallJoker Thanks for that additional info. Yes, I figured that it wasn't the intent of the recent changes, and was guessing that the suggestion would not be viewed favorably, at least not in any official capacity. However, it hadn't been explicitly discussed and rejected, so no real sense of it being "bad" vs "not ideal but passible if need be".

It sounds good that there isn't a ton of work left to get to this goal. But, for now, at least `get_mapgen_setting()` get's us pretty close. And, it'll be easy enough to swap out for the ideal method whenever that becomes available.

**Comment by ResuUman** [comment]
Suggestion: global settings should be protected from (unintentional,all?) changes (by game,by player?) during play.

example: game Castrum quietly uses: minetest.setting_set("time_speed", "0") and not returning original value (eg. 72) after quit.

**Comment by dibesfer** [comment]
It'd be useful for adding server info, server_address and so on, on a specific world meant for server.

**Comment by NathanSalapat** [comment]
You can declare a .conf file when running a sever via the command line. This feature would be very nice to have, but for your specific use case there is already an option.
